python/src/uagents/registration.py

```diff
@@ -23,6 +24,51 @@
 from uagents.types import AgentEndpoint
 
 
+class VerifiableModel(BaseModel):
+    agent_address: str
+    signature: Optional[str] = None
+    timestamp: Optional[int] = None
+
+    def sign(self, identity: Identity):
+        self.timestamp = int(time.time())
+        digest = self._build_digest()
+        self.signature = identity.sign_digest(digest)
+
+    def verify(self) -> bool:
+        return self.signature is not None and Identity.verify_digest(
+            self.agent_address, self._build_digest(), self.signature
+        )
+
+    def _build_digest(self) -> bytes:
+        sha256 = hashlib.sha256()
+        sha256.update(
+            json.dumps(
+                self.model_dump(exclude={"signature"}),
+                sort_keys=True,
+                separators=(",", ":"),
+            ).encode("utf-8")
+        )
+        return sha256.digest()
+
+
+class AgentRegistrationAttestation(VerifiableModel):
+    protocols: List[str]
+    endpoints: List[AgentEndpoint]
+    metadata: Optional[Dict[str, Union[str, Dict[str, str]]]] = None
+
+    @field_serializer("protocols")
```

Comment:
Is there a way to keep the original object unchanged and only apply these to a copy with will be used for generating the digest?

Reply:
I can think of a couple but do we need to keep these separated? The order of the data has no impact on the following processing as far as I’m aware and this method provides some consistency as every time the model is dumped it produces the same result.

Reply:
Okay, at first I was thinking that there could be some functional reason for ordering the endpoints, but probably that only applies to the agent itself and not the attestation, so I think we can keep these field serializers.

Reply:
I created these serializers because in the original digest function the protocols and endpoints were ordered. Looking at it now it makes no sense they exist in the first place.

The connotation of the digest is inherently different from that of a manifest where we give leeway with the order of the fields etc. by always ordering before calculating the digest. Here, it's just a means of creating a verification method for integrity. Also, in the endpoint we actually remove all related data first before taking it 1:1 from the attestation so there is no ordering necessary here as well.

For that reason I'll remove them.